### PR TITLE
Fix VarlenEntry CompareEqualOrNot to ignore reclaim bit properly.

### LINKED ISSUE #1065: Fail to Compare VARCHARs with More than 12 Characters
12a's as the predicate works but get nothing with 13 a's. How to reproduce:
```
CREATE TABLE supply (s_suppkey INT, s_comment VARCHAR);
INSERT INTO supply VALUES (1, 'aaaaaaaaaaaa');
SELECT s_suppkey from supply WHERE s_comment = 'aaaaaaaaaaaa';
 s_suppkey
-----------
         1
(1 row)
INSERT INTO supply VALUES (1, 'aaaaaaaaaaaaa');
SELECT s_suppkey from supply WHERE s_comment = 'aaaaaaaaaaaaa';
 s_suppkey
-----------
(0 rows)
```

### PR DESCRIPTION
**Pull request by lmwnshn**
Fix #1065.

Because of endianness-related issues, the reclaim bit is not
located at the first bit. Therefore the original memcmp-based
approach was failing. The code now reads a uint64_t and masks
out the relevant bit.

@dniu16 What's the current status of the JUnit-based test files?
How should tests be added? Thanks!

We want to add the following test case:
```
create table foo (a varchar);
insert into foo values ('11223344556677889900112233');
select a from foo where a = '11223344556677889900112233';
```

### CODE REVIEW

File: src/include/storage/storage_defs.h [hunk 1]
```diff
@@ -416,30 +416,43 @@ class VarlenEntry {
...
-    bool following_bytes_same =
-        0 == std::memcmp(reinterpret_cast<const char *>(&left) + 1, reinterpret_cast<const char *>(&right) + 1,
-                         sizeof(left.size_) + PrefixSize() - 1);
-    if (first_byte_same && following_bytes_same) {
-      // Prefix and length are equal.
+    const uint64_t left_size_prefix = *reinterpret_cast<const uint64_t *>(&left);
+    const uint64_t right_size_prefix = *reinterpret_cast<const uint64_t *>(&right);
+    // Mask off the reclaimability bit and compare.
+    // Note that due to endianness, when reading this out as a uint64_t the reclaim bit is NOT the sign bit.
+    constexpr uint64_t remove_reclaim_mask = 0xffffffff7fffffff;  // You're just gonna have to believe me
```
Comment (pmenon) [inline src/include/storage/storage_defs.h:424]: This seems incredible unintuitive - you got it wrong in an earlier commit.

If we're going to stick with this, I'd like to see a performance benchmark against a simpler version that checks the 7 bytes after the first as a "lossy" filter, then validate using the actual size. I think this is easier to understand and will probably have equivalent performance without the need for magic numbers.

**Comment by dniu16** [comment]
> Fix #1065.
> 
> Because of endianness-related issues, the reclaim bit is not
> located at the first bit. Therefore the original memcmp-based
> approach was failing. The code now reads a uint64_t and masks
> out the relevant bit.
> 
> @dniu16 What's the current status of the JUnit-based test files?
> How should tests be added? Thanks!
> 
> We want to add the following test case:
> 
> ```
> create table foo (a varchar);
> insert into foo values ('11223344556677889900112233');
> select a from foo where a = '11223344556677889900112233';
> ```

Sure I'll add this test case to the select.test file. (FYI instruction for adding tests is located under junit/README.md)

File: src/include/storage/storage_defs.h (hunk 1 above)
Comment (lmwnshn) [reply src/include/storage/storage_defs.h:424]: Per discussion in meeting, leaning towards checking size followed by prefix separately. It is not possible to check the 7 bytes together.

**Comment by mbutrovich** [comment]
Didn't see a meaningful performance difference between this approach and the previous masking approach, so let's go with this.

**Comment by mbutrovich** [comment]
I re-ran numbers on the NUC this morning, and it doesn't match what I saw on my laptop. On the laptop it was basically a wash and within 5%. On the NUC today, however:

The previous approach using a single comparison of length and prefix:
```
---------------------------------------------------------------------------------------------------------
Benchmark                                                                  Time           CPU Iterations
---------------------------------------------------------------------------------------------------------
VarlenEntryBenchmark/EqualityInlineEqualPrefixOnly                         1 ns          1 ns  911948518   1.55256G items/s
VarlenEntryBenchmark/EqualityInlineEqualPrefixDifferentLength              0 ns          0 ns 1000000000   2.71711G items/s
VarlenEntryBenchmark/EqualityInlineEqualPrefixEqualLength                  3 ns          3 ns  229347045   316.641M items/s
VarlenEntryBenchmark/EqualityInlineDifferentPrefixEqualLength              0 ns          0 ns 1000000000    2.7868G items/s
VarlenEntryBenchmark/EqualityNotInlineEqualContentEqualLength              5 ns          5 ns  155228937    211.45M items/s
VarlenEntryBenchmark/EqualityNotInlineDifferentContentEqualLength          5 ns          5 ns  139108238   189.656M items/s
Process finished with exit code 0
```
As the PR currently stands (compare size, compare prefix):
```
---------------------------------------------------------------------------------------------------------
Benchmark                                                                  Time           CPU Iterations
---------------------------------------------------------------------------------------------------------
VarlenEntryBenchmark/EqualityInlineEqualPrefixOnly                         1 ns          1 ns  640875816   927.395M items/s
VarlenEntryBenchmark/EqualityInlineEqualPrefixDifferentLength              1 ns          1 ns 1000000000   1.82181G items/s
VarlenEntryBenchmark/EqualityInlineEqualPrefixEqualLength                  3 ns          3 ns  232724057   317.064M items/s
VarlenEntryBenchmark/EqualityInlineDifferentPrefixEqualLength              1 ns          1 ns 1000000000   1.59242G items/s
VarlenEntryBenchmark/EqualityNotInlineEqualContentEqualLength              4 ns          4 ns  161897879   220.564M items/s
VarlenEntryBenchmark/EqualityNotInlineDifferentContentEqualLength          5 ns          5 ns  155991465   200.439M items/s
Process finished with exit code 0
```
The benchmarks `EqualityInlineEqualPrefixDifferentLength` and `EqualityInlineDifferentPrefixEqualLength` show a pronounced difference.

**Comment by mbutrovich** [comment]
For reference, the laptop (Clang 8, macOS 10.15.6, release) numbers:

The previous approach using a single comparison of length and prefix:
```
---------------------------------------------------------------------------------------------------------
Benchmark                                                                  Time           CPU Iterations
---------------------------------------------------------------------------------------------------------
VarlenEntryBenchmark/EqualityInlineEqualPrefixOnly                         0 ns          0 ns 1000000000    3.0744G items/s
VarlenEntryBenchmark/EqualityInlineEqualPrefixDifferentLength              0 ns          0 ns 1000000000   3.69953G items/s
VarlenEntryBenchmark/EqualityInlineEqualPrefixEqualLength                  0 ns          0 ns 1000000000   3.72456G items/s
VarlenEntryBenchmark/EqualityInlineDifferentPrefixEqualLength              0 ns          0 ns 1000000000   3.65038G items/s
VarlenEntryBenchmark/EqualityNotInlineEqualContentEqualLength              0 ns          0 ns 1000000000   3.62827G items/s
VarlenEntryBenchmark/EqualityNotInlineDifferentContentEqualLength          0 ns          0 ns 1000000000   3.73108G items/s
Process finished with exit code 0
```
As the PR currently stands (compare size, compare prefix) based on discussion in the standup:
```
---------------------------------------------------------------------------------------------------------
Benchmark                                                                  Time           CPU Iterations
---------------------------------------------------------------------------------------------------------
VarlenEntryBenchmark/EqualityInlineEqualPrefixOnly                         0 ns          0 ns 1000000000   3.07964G items/s
VarlenEntryBenchmark/EqualityInlineEqualPrefixDifferentLength              0 ns          0 ns 1000000000   3.73849G items/s
VarlenEntryBenchmark/EqualityInlineEqualPrefixEqualLength                  0 ns          0 ns 1000000000   3.58938G items/s
VarlenEntryBenchmark/EqualityInlineDifferentPrefixEqualLength              0 ns          0 ns 1000000000   3.73986G items/s
VarlenEntryBenchmark/EqualityNotInlineEqualContentEqualLength              0 ns          0 ns 1000000000   3.72723G items/s
VarlenEntryBenchmark/EqualityNotInlineDifferentContentEqualLength          0 ns          0 ns 1000000000   3.74296G items/s
Process finished with exit code 0
```
At this throughput performance fluctuates 5-10% run to run so I couldn't see a meaningful difference between the two approaches.

**Comment by pmenon** [comment]
Interesting benchmark results! I'd say the perf benefits warrant the optimization. Use the previous size+prefix single-comparison version, but add a comment explaining the origin of the bitmask (i.e., big-endian layout and location of the reclaim bit, assumption of data type for `size` value, etc.). Otherwise, LGTM.

**Comment by mbutrovich** [comment]
Sounds good. I'll fix it up tomorrow.

**Pull request merged**